File: python/example_tts_low_latency_http.py
#!/usr/bin/env python3
"""
Example script for low-latency TTS synthesis using HTTP streaming.

This script demonstrates how to achieve the lowest possible time-to-first-byte (TTFB)
with HTTP streaming by warming up the connection before timing synthesis.

Key technique: Use a persistent session to pre-establish the TCP+TLS connection
with a small warmup request, then measure only the synthesis latency.
"""

import base64
import json
import os
import time

import requests


def check_api_key():
    """Check if INWORLD_API_KEY environment variable is set."""
    api_key = os.getenv("INWORLD_API_KEY")
    if not api_key:
        print("Error: INWORLD_API_KEY environment variable is not set.")
        print("Please set it with: export INWORLD_API_KEY=your_api_key_here")
        return None
    return api_key


def http_streaming_tts(api_key, text, voice_id, model_id):
    """
    Measure low-latency TTS using HTTP streaming with connection warmup.

    Uses a persistent session to pre-establish the TCP+TLS connection,
    then measures TTFB from the actual synthesis request only.

    Args:
        api_key: API key for authentication
        text: Text to synthesize
        voice_id: Voice ID to use
        model_id: Model ID to use

    Returns:
        dict: Latency metrics (ttfb, total_time, audio_bytes) or None on error
    """
    url = "https://api.inworld.ai/tts/v1/voice:stream"

    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Basic {api_key}"
    }

    request_data = {
        "text": text,
        "voice_id": voice_id,
        "model_id": model_id,
        "audio_config": {
            "audio_encoding": "OGG_OPUS",
            "sample_rate_hertz": 24000,
            "bit_rate": 32000
        }
    }

    # Use a session for connection reuse (TCP+TLS keepalive)
    session = requests.Session()
    session.headers.update(headers)

    try:
        # Warmup: establish TCP+TLS connection before timing
        warmup_data = {
            "text": "hi",
            "voice_id": voice_id,
            "model_id": model_id,
            "audio_config": {
                "audio_encoding": "OGG_OPUS",
                "sample_rate_hertz": 24000,
                "bit_rate": 32000
            }
        }

        # Non-streaming so the full response is read and the connection
        # is returned to the pool for reuse by the timed request.
        session.post(url, json=warmup_data)

        # Start timer - connection already established
        start_time = time.time()
        ttfb = None
        total_audio_bytes = 0

        with session.post(url, json=request_data, stream=True) as response:
            response.raise_for_status()

            for line in response.iter_lines(decode_unicode=True):
                if line.strip():
                    try:
                        chunk_data = json.loads(line)
                        result = chunk_data.get("result")
                        if result and "audioContent" in result:
                            audio_chunk = base64.b64decode(result["audioContent"])
                            if ttfb is None:
                                ttfb = time.time() - start_time
                            total_audio_bytes += len(audio_chunk)
                    except json.JSONDecodeError:
                        continue

        total_time = time.time() - start_time
        return {"ttfb": ttfb, "total_time": total_time, "audio_bytes": total_audio_bytes}

    except requests.exceptions.RequestException as e:
        print(f"HTTP Error: {e}")
        if hasattr(e, 'response') and e.response is not None:
            try:
                error_detail = e.response.json()
                print(f"   Error details: {error_detail}")
            except:
                print(f"   Response text: {e.response.text}")
        return None
    finally:
        session.close()


def main():
    """Main function to demonstrate low-latency HTTP streaming TTS."""
    print("Inworld TTS Low-Latency HTTP Streaming")
    print("=" * 45)

    # Check API key
    api_key = check_api_key()
    if not api_key:
        return 1

    # Configuration
    text = "Life moves pretty fast. Look around once in a while, or you might miss it."
    voice_id = "Dennis"
    model_id = "inworld-tts-1.5-mini"

    print(f"   Text: \"{text}\"")
    print(f"  Voice: {voice_id}")
    print(f"  Model: {model_id}")
    print(f"\nWarming up connection, then generating audio...\n")

    try:
        result = http_streaming_tts(api_key, text, voice_id, model_id)

        if result:
            print(f"TTFB:         {result['ttfb']*1000:.1f} ms")
            print(f"Total time:   {result['total_time']*1000:.1f} ms")
            print(f"Audio bytes:  {result['audio_bytes']}")
        else:
            print("Synthesis failed.")
            return 1

    except Exception as e:
        print(f"\nLatency test failed: {e}")
        return 1

    return 0


if __name__ == "__main__":
    exit(main())

File: js/example_tts_low_latency_http.js
#!/usr/bin/env node
/**
 * Example script for low-latency TTS synthesis using HTTP streaming.
 *
 * This script demonstrates how to achieve the lowest possible time-to-first-byte (TTFB)
 * with HTTP streaming by warming up the connection before timing synthesis.
 *
 * Key technique: Use a warmup request to pre-establish the TCP+TLS connection,
 * then measure only the synthesis latency on the reused connection.
 */

/**
 * Check if INWORLD_API_KEY environment variable is set.
 * @returns {string|null} API key or null if not set
 */
function checkApiKey() {
    const apiKey = process.env.INWORLD_API_KEY;
    if (!apiKey) {
        console.log('Error: INWORLD_API_KEY environment variable is not set.');
        console.log('Please set it with: export INWORLD_API_KEY=your_api_key_here');
        return null;
    }
    return apiKey;
}

/**
 * Measure low-latency TTS using HTTP streaming with connection warmup.
 *
 * Uses a warmup request to pre-establish the TCP+TLS connection,
 * then measures TTFB from the actual synthesis request only.
 *
 * @param {string} apiKey - API key for authentication
 * @param {string} text - Text to synthesize
 * @param {string} voiceId - Voice ID to use
 * @param {string} modelId - Model ID to use
 * @returns {Object|null} Latency metrics {ttfb, totalTime, audioBytes} or null on error
 */
async function httpStreamingTts(apiKey, text, voiceId, modelId) {
    const url = 'https://api.inworld.ai/tts/v1/voice:stream';

    const headers = {
        'Content-Type': 'application/json',
        'Authorization': `Basic ${apiKey}`,
        'Connection': 'keep-alive'
    };

    const requestData = {
        text: text,
        voice_id: voiceId,
        model_id: modelId,
        audio_config: {
            audio_encoding: 'OGG_OPUS',
            sample_rate_hertz: 24000,
            bit_rate: 32000
        }
    };

    try {
        // Warmup: establish TCP+TLS connection before timing
        const warmupData = {
            text: 'hi',
            voice_id: voiceId,
            model_id: modelId,
            audio_config: {
                audio_encoding: 'OGG_OPUS',
                sample_rate_hertz: 24000,
                bit_rate: 32000
            }
        };

        const warmupResponse = await fetch(url, {
            method: 'POST',
            headers: headers,
            body: JSON.stringify(warmupData)
        });

        // Fully consume warmup response so the connection is returned to the pool
        await warmupResponse.arrayBuffer();

        // Start timer - connection already established
        const startTime = Date.now();
        let ttfb = null;
        let totalAudioBytes = 0;

        const response = await fetch(url, {
            method: 'POST',
            headers: headers,
            body: JSON.stringify(requestData)
        });

        if (!response.ok) {
            let errorDetails = '';
            try {
                errorDetails = await response.text();
            } catch {}
            console.log(`HTTP Error: ${response.status} - ${errorDetails}`);
            return null;
        }

        const reader = response.body.getReader();
        const decoder = new TextDecoder();
        let buffer = '';

        while (true) {
            const { done, value } = await reader.read();
            if (done) break;

            buffer += decoder.decode(value, { stream: true });
            const lines = buffer.split('\n');
            buffer = lines.pop() || '';

            for (const line of lines) {
                if (line.trim()) {
                    try {
                        const chunkData = JSON.parse(line);
                        const result = chunkData.result;
                        if (result && result.audioContent) {
                            const audioChunk = Buffer.from(result.audioContent, 'base64');
                            if (ttfb === null) {
                                ttfb = (Date.now() - startTime) / 1000;
                            }
                            totalAudioBytes += audioChunk.length;
                        }
                    } catch {
                        continue;
                    }
                }
            }
        }

        const totalTime = (Date.now() - startTime) / 1000;
        return { ttfb, totalTime, audioBytes: totalAudioBytes };

    } catch (error) {
        console.log(`HTTP Error: ${error.message}`);
        return null;
    }
}

/**
 * Main function to demonstrate low-latency HTTP streaming TTS.
 */
async function main() {
    console.log('Inworld TTS Low-Latency HTTP Streaming');
    console.log('=' + '='.repeat(44));

    // Check API key
    const apiKey = checkApiKey();
    if (!apiKey) {
        return 1;
    }

    // Configuration
    const text = "Life moves pretty fast. Look around once in a while, or you might miss it.";
    const voiceId = 'Dennis';
    const modelId = 'inworld-tts-1.5-mini';

    console.log(`   Text: "${text}"`);
    console.log(`  Voice: ${voiceId}`);
    console.log(`  Model: ${modelId}`);
    console.log('\nWarming up connection, then generating audio...\n');

    try {
        const result = await httpStreamingTts(apiKey, text, voiceId, modelId);

        if (result) {
            console.log(`TTFB:         ${(result.ttfb * 1000).toFixed(1)} ms`);
            console.log(`Total time:   ${(result.totalTime * 1000).toFixed(1)} ms`);
            console.log(`Audio bytes:  ${result.audioBytes}`);
        } else {
            console.log('Synthesis failed.');
            return 1;
        }

    } catch (error) {
        console.log(`\nLatency test failed: ${error.message}`);
        return 1;
    }

    return 0;
}

if (require.main === module) {
    main().then(process.exit);
}
